woocommerce/woocommerce-android
# Configure JaCoCo

Closes: #6048 


Internal communication links: p91TBi-7v0-p2 p91TBi-7Sy-p2

cc: @thehenrybyrd 

### Description
This PR integrates JaCoCo - a tool that will allow us to measure code coverage and make tasks like #6047 easier.

This is more of an experiment for now to see how things work. We do not set any limits for code coverage or expectations. Also JaCoCo won't interfere with any of the existing tools/processes - it's a safe experiment.

#### Why not other tool, e.g. [Kover](https://github.com/Kotlin/kotlinx-kover)?
1. Requires more configuration (e.g. filtering out specific build types/flavours. See [issues](https://github.com/Kotlin/kotlinx-kover/issues?q=is%3Aissue+is%3Aopen+android))
2. Kover is not supported by a tool that can help us measure coverage over time - CodeCov.

#### Why not using code coverage results from Android Studio?
While it's possible, it has some flaws. E.g. we can't get those results without Android Studio/IntelliJ so it's not possible to get results on CI (and then send it to CodeCov). Another thing is that exclusion rules would've been standardised between Android Studio configurations on each team member machine. It adds not worth effort IMO.

### Testing instructions
1. Please run `./gradlew jacocoTestReport`
2. Open html report at `WooCommerce/build/reports/jacoco/wasabiDebug/index.html`
3. Browse it. Make sure that there's no sign of generated code.
4. Assert that code from `cardreader` module is also presented in report (package `com.woocommerce.android.cardreader`)

### Images/gif

<img width="1292" alt="image" src="https://user-images.githubusercontent.com/5845095/158989798-15a57d79-6494-4764-9d0c-5fb270d8c7a9.png">
(...)
<img width="1277" alt="image" src="https://user-images.githubusercontent.com/5845095/158989843-a9729f6c-e254-45e7-bdcc-109592dd8a12.png">
